auth/downscoping/snippets.py

```diff
@@ -0,0 +1,145 @@
+            to ensure access is restricted to only objects starting with this
+            prefix string.
+
+    Returns:
+        Tuple[str, datetime.datetime]: The downscoped access token and its expiry date.
+    """
+    # [START auth_downscoping_rules]
+    # Initialize the credential access boundary rules.
+    available_resource = f"//storage.googleapis.com/projects/_/buckets/{bucket_name}"
+    # Downscoped credentials will have readonly access to the resource.
+    available_permissions = ["inRole:roles/storage.objectViewer"]
+    # Only objects starting with the specified prefix string in the object name
+    # will be allowed read access.
+    availability_expression = (
+        "resource.name.startsWith('projects/_/buckets/{}/objects/{}')".format(
+            bucket_name, object_prefix
+        )
+    )
+    availability_condition = downscoped.AvailabilityCondition(availability_expression)
+    # Define the single access boundary rule using the above properties.
+    rule = downscoped.AccessBoundaryRule(
+        available_resource=available_resource,
+        available_permissions=available_permissions,
+        availability_condition=availability_condition,
+    )
+    # Define the credential access boundary with all the relevants rules.
+    credential_access_boundary = downscoped.CredentialAccessBoundary(rules=[rule])
+    # [END auth_downscoping_rules]
+
+    # [START auth_downscoping_initialize_downscoped_cred]
+    # Retrieve the source credentials via ADC.
+    source_credentials, _ = google.auth.default()
+    if source_credentials.requires_scopes:
+        source_credentials = source_credentials.with_scopes(
+            ["https://www.googleapis.com/auth/cloud-platform"]
+        )
+
+    # Create the downscoped credentials.
+    downscoped_credentials = downscoped.Credentials(
+        source_credentials=source_credentials,
+        credential_access_boundary=credential_access_boundary,
+    )
+
+    # Refresh the tokens.
+    downscoped_credentials.refresh(requests.Request())
+
+    # These values will need to be passed to the Token Consumer.
+    access_token = downscoped_credentials.token
+    expiry = downscoped_credentials.expiry
+    # [END auth_downscoping_initialize_downscoped_cred]
+    return (access_token, expiry)
+
+
+# [END auth_downscoping_token_broker]
+
+
+# [START auth_downscoping_token_consumer]
+def token_consumer(bucket_name, object_name):
+    """Tests token consumer readonly access to the specified object.
+
+    Args:
+        bucket_name (str): The name of the cloud storage bucket.
+        object_name (str): The name of the object in the storage bucket to
+            read.
+    """
+    # Create the OAuth credentials from the downscoped token and pass a
+    # refresh handler to handle token expiration. Passing the original
+    # downscoped token or the expiry here is optional, as the refresh_handler
+    # will generate the downscoped token on demand.
```

Comment:
I'm a little fuzzy on what this means. We say this is "optional," but I think what we really mean is "we're not passing in the token/expiry in this example, because for demonstration purposes, we generate the token locally." True?

Also, should this say "...will generate a new downscoped token..."?

Reply:
I agree this is confusing. I have rephrased to explain why we are passing a `refresh_handler`. I think this is better.